modules/globallock/lock.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package globallock
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/nosql"
+	"code.gitea.io/gitea/modules/setting"
+
+	redsync "github.com/go-redsync/redsync/v4"
+	goredis "github.com/go-redsync/redsync/v4/redis/goredis/v9"
+)
+
+type Locker interface {
+	Lock() error
+	TryLock() (bool, error)
+	Unlock() (bool, error)
+}
+
+type LockService interface {
+	GetLock(name string) Locker
```

Comment:
It also needs comments to clarify the behaviors for some edge cases.

For example, `GetLock` shouldn't used for temporary tasks, like `GetLock(fmt.Sprintf("migration_task_%d", task.ID)`, since `memoryLockService` will store the lock forever without GC even it has been unlocked.